doc/03-omnifest/02-external.md

```diff
@@ -1,22 +1,117 @@
 # External
 
-## osbuild
+External directives are directives that are implemented externally from `otk`.
+They are meant to be used to provide target-specific behavior and can be used
+to express things where `otk` is not expressive enough.
+
+## Protocol
+
+Directives are small executables that receive JSON and are expected to output
+JSON again in a specific format. Additional keys are not allowed.
+
+When we have an [omnifest](./index.md) that looks like this:
+
+```yaml
+otk.external.name:
+  child:
+    - 1
+  options: "are here"
+```
+
+The external gets called with the following JSON:
+
+```json
+{
+  "tree": {
+    "otk.external.name": {
+      "child": [1],
+      "options": "are here"
+    }
+  }
+}
+```
+
+And is expected to return the following JSON:
+
+```json
+{
+  "tree": {
+    "what": {
+      "you": "want"
+    }
+  }
+}
+```
+
+Which will lead to the following YAML structure:
+
+```yaml
+tree:
+  what:
+    you: "want"
+```
+
+If the external returns `{}`, an empty object, `otk` will assume that there is
+no tree to replace and remove the node instead. This will turn the following
+YAML:
+
+```yaml
+dummy:
+  otk.external.name:
+    options:
+```
+
+Into this YAML structure:
+
+```yaml
+dummy:
+```
+
+## Paths
+
+`otk` will look for external directives in the following paths, stopping when
+it finds the first match:
+
+- A path defined by the `OTK_EXTERNAL_PATH` environment variable.
+- `/usr/local/libexec/otk`
+- `/usr/libexec/otk`
+- `/usr/local/lib/otk`
+- `/usr/lib/otk`
+
+The filename for an external executable is based on the external name. When the
+following directive is encountered: `otk.external.<name>.<argument>` then
```

Comment:
Do you have an example for `<argument>` ? I'm not sure I follow why we need this mapping. As a stawman, maybe something very simple. The search path is the same as above but contains "/external/" as the last component (so /usr/libexec/otk/external" and then just `osbuild_dnf4" ? Or are we limiting ourselfs here is do something very simple like this?

Reply:
Is this meant to be like how git (and others) work with subcommands?

Reply:
I think in the interest of keeping things simple for the initial implementation, we can avoid adding features that can be expressed easily with other features, especially if the alternative isn't too complicated.  A subcommand here can easily be expressed using an option for the external.  

```yaml
...
  otk.external.myscript.subcommand_1:
    option: "value"
```

is not that much cleaner than:
```yaml
...
  otk.external.myscript:
    command: "subcommand_1"
    options: "value"
```

I think the fact that the first form requires an extra section in the specification and docs is a sign that it's not necessary given how the alternative isn't really that complicated.

Reply:
Done.